modules/webgl/src/adapter/resources/webgl-transform-feedback.ts

```diff
@@ -0,0 +1,253 @@
+import type {Device, ShaderLayout, TransformFeedbackProps} from '@luma.gl/core';
+import {log, isObjectEmpty, TransformFeedback, Buffer} from '@luma.gl/core';
+import {GL} from '@luma.gl/constants';
+import {WebGLDevice} from '../webgl-device';
+import {WEBGLBuffer, isWebGL2} from '../..';
+
+/** For bindRange */
+type BufferRange = {
+  buffer: WEBGLBuffer;
+  byteOffset?: number;
+  byteLength?: number;
+};
+
+export class WEBGLTransformFeedback extends TransformFeedback {
+  get [Symbol.toStringTag](): string {
+    return 'TransformFeedback';
+  }
+
+  readonly device: WebGLDevice;
+  readonly gl2: WebGL2RenderingContext;
+  readonly handle: WebGLTransformFeedback;
+
+  /**
+   * NOTE: The Model already has this information while drawing, but
+   * TransformFeedback currently needs it internally, to look up
+   * varying information outside of a draw() call.
+   */
+  readonly layout: ShaderLayout;
+  buffers: Record<string, BufferRange> = {};
+  unusedBuffers: Record<string, Buffer> = {};
+  // NOTE: The `bindOnUse` flag is a major workaround:
+  // See https://github.com/KhronosGroup/WebGL/issues/2346
+  bindOnUse = true;
+  private _bound: boolean = false;
+
+  constructor(device: WebGLDevice, props: TransformFeedbackProps) {
+    super(device, props);
+
+    device.assertWebGL2();
+    this.device = device;
+    this.gl2 = device.gl2;
+    this.handle = this.props.handle || this.gl2.createTransformFeedback();
+    this.layout = this.props.layout;
+
+    // TODO(v9): Should this logic exist in the constructor?
+    try {
+      this.gl2.bindTransformFeedback(GL.TRANSFORM_FEEDBACK, this.handle);
+      for (const [locationOrName, bufferOrRange] of Object.entries(this.props.buffers)) {
+        const location = this._getVaryingIndex(locationOrName);
+        const range = this._getBufferRange(bufferOrRange as WEBGLBuffer | BufferRange);
+        if (location < 0) {
+          this.unusedBuffers[locationOrName] = range.buffer;
+          log.warn(`${this.id} unused varying buffer ${locationOrName}`)();
+          return this;
+        }
+        this.buffers[location] = range;
+      }
+    } finally {
+      this.gl2.bindTransformFeedback(GL.TRANSFORM_FEEDBACK, null);
+    }
```

Comment:
I don't think we need this chunk of code here in the constructor, and could instead replace it with something like:

```suggestion
    if (this.props.buffers) {
        this.setBuffers(this.props.buffers);
    }
```

Assuming tests pass does that sound reasonable? Or is there another motive for the code?

Reply:
I would say that generally we aim for an immutable GPU resource API to align with spirit of WebGPU, and in such an API model, the user can count on fields being initialized by the constructor. Also in typescript, this means that fields can be typed without optionality (e.g. `Buffer | null`) since they are guaranteed to be initialized to a valid object in the constructor.

Depending on the usage pattern of TransformFeedback it could be a tradeoff, but generally I feel it is a slightly more elegant to do initialization in the constructor.

Reply:
Agreed on immutable API, and initializing during the constructor makes sense to me! My concern is that this code is duplicated, and we could call `setBuffers(this.prop.buffers)` to get the same result. I want ahead and refactored that part.